modules/onewire/onewire.c

```diff
@@ -0,0 +1,418 @@
+    return 1;
+}
+
+// ONEWIRE IMPLEMENTATION
+
+uint8_t OneWire_Init(OneWire_t* OneWireStruct, uint32_t PalLine) {
+    if (!OneWireStruct) return 0;
+    OneWireStruct->PalLine = PalLine;
+    return 1;
+}
+
+char OneWire_Reset(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return -1;
+    uint8_t i;
+
+    /* Line low, and wait */
+    ONEWIRE_OUTPUT(OneWireStruct);
+    ONEWIRE_LOW(OneWireStruct);
+    ONEWIRE_DELAY(ONEWIRE_TX_MIN_RESET_PULSE_TIME_USEC);
+    
+    /* Release line and wait */
+    ONEWIRE_INPUT(OneWireStruct);
+    ONEWIRE_DELAY(ONEWIRE_WAIT_PRESENCE_PULSE_TIME_USEC);
+    
+    /* Check bit value */
+    i = palReadLine(OneWireStruct->PalLine);
+
+    /* Complete reset pulse min time */
+    ONEWIRE_DELAY(ONEWIRE_RX_MIN_RESET_PULSE_TIME_USEC);
+    /* Return value of presence pulse, 0 = OK, 1 = ERROR */
+    return i;
+}
+
+uint8_t OneWire_WriteBit(OneWire_t* OneWireStruct, uint8_t bit) {
+    if (!OneWireStruct) return 0;
+    if (bit) {
+        /* Set line low */
+        ONEWIRE_OUTPUT(OneWireStruct);
+        ONEWIRE_LOW(OneWireStruct);
+        ONEWIRE_DELAY(ONEWIRE_TX_WRITE_1_BIT_LO_TIME_USEC);
+
+        /* release line, Hold for remainder of min time for write 1 slot */
+        ONEWIRE_INPUT(OneWireStruct);
+        ONEWIRE_DELAY(ONEWIRE_TX_WRITE_1_BIT_HI_TIME_USEC);
+    } else {
+        /* Set line low */
+        ONEWIRE_OUTPUT(OneWireStruct);
+        ONEWIRE_LOW(OneWireStruct);
+        ONEWIRE_DELAY(ONEWIRE_TX_WRITE_0_BIT_LO_TIME_USEC);  
+
+        /* release line, wait for nonzero recovery time */
+        ONEWIRE_INPUT(OneWireStruct);
+        ONEWIRE_DELAY(ONEWIRE_TX_RECOVER_TIME_USEC);
+    }
+    return 1;
+}
+
+char OneWire_ReadBit(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return -1;
+    uint8_t bit = 0;
+
+    /* Line low */
+    ONEWIRE_OUTPUT(OneWireStruct);
+    ONEWIRE_LOW(OneWireStruct);
+    ONEWIRE_DELAY(ONEWIRE_RX_READ_BIT_LO_TIME_USEC); 
+
+    /* Release line */
+    ONEWIRE_INPUT(OneWireStruct);
+    ONEWIRE_DELAY(ONEWIRE_RX_READ_BIT_WAIT_BEFORE_SAMPLE_TIME_USEC); // 
+
+    /* Read line value near end of 15us */
+    if (palReadLine(OneWireStruct->PalLine)) bit = 1;
+
+    /* Wait 50us to complete minimum 60us period */
+    ONEWIRE_DELAY(50);
+
+    /* Return bit value */
+    return bit;
+}
+
+uint8_t OneWire_WriteByte(OneWire_t* OneWireStruct, uint8_t byte) {
+    if (!OneWireStruct) return 0;
+    uint8_t i = 8;
+    /* Write 8 bits */
+    while (i--) {
+        /* LSB bit is first */
+        OneWire_WriteBit(OneWireStruct, byte & 0x01);
+        byte >>= 1;
+    }
+    return 1;
+}
+
+char OneWire_ReadByte(OneWire_t* OneWireStruct) {
+    if (!OneWireStruct) return -1;
+    uint8_t i = 8, byte = 0;
+    while (i--) {
+        /* data is transferred LSB first, shift summed byte first by 1,
+        then shift data byte by 7 to put in appropriate power */
+        byte >>= 1;
+        byte |= (OneWire_ReadBit(OneWireStruct) << 7);
+    }

```

Comment:
For this function to work `outval` should be initialized with value of 0, otherwise, the interpreted readbyte value would be interpreted incorrectly. But yes I agree.